Src/Particle/AMReX_ParticleContainerI.H

```diff
@@ -701,87 +722,142 @@ ParticleContainer<NStructReal, NStructInt, NArrayReal, NArrayInt, Allocator>
-                        const Real           total_mass = old_mass + new_mass;
-                        //
-                        // Set the position to the center of mass.
-                        //
-                        for (int i = 0; i < AMREX_SPACEDIM; i++)
-                        {
-                            pold.pos(i) = static_cast<ParticleReal>((old_mass*pold.pos(i) + new_mass*pnew.pos(i))/total_mass);
-                        }
-                        AMREX_ASSERT(this->Index(pold, level) == cell);
-                        //
-                        // Set the metadata (presumably velocity) to the mass-weighted average.
-                        //
-                        for (int i = 1; i < NStructReal; i++)
-                        {
-                            pold.rdata(i) = static_cast<ParticleReal>((old_mass*pold.rdata(i) + new_mass*pnew.rdata(i))/total_mass);
-                        }
-                        pold.rdata(0) = static_cast<ParticleReal>(total_mass);
-                    }
-                }
+               //Ordering may make this not unique
+               for (int i = 0; i < NArrayReal; ++i)
+                 amrex::Gpu::Atomic::AddNoRet(&partData(iv,AMREX_SPACEDIM+NStructReal+i), partData(iv,AMREX_SPACEDIM)!=0.0 ? static_cast<Real>(0) : static_cast<Real>(p.rdata(NStructReal+i)));
+
+               for (int i = 0; i < AMREX_SPACEDIM; ++i)
+               {
+                   amrex::Gpu::Atomic::AddNoRet(&partData(iv,i), static_cast<Real>((p.rdata(0)*p.pos(i))));
+               }
+               for (int i = 1; i < NStructReal; ++i)
+               {
+                   amrex::Gpu::Atomic::AddNoRet(&partData(iv,AMREX_SPACEDIM+i), static_cast<Real>((p.rdata(0)*p.rdata(i))));
+               }
+               for (int i = 0; i < 1; ++i)
+                   amrex::Gpu::Atomic::AddNoRet(&partData(iv,AMREX_SPACEDIM+i), static_cast<Real>(p.rdata(0)));
             }
 
-            //
-            // Add the aggregated particles to the virtuals.
-            //
-            for (const auto& agg_particle : agg_map)
+        }); //skipping extra false argument, doing mf.setVal(0) at beginning
+
+        amrex::ParticleToMesh(*this, imf, level,
+        [=] AMREX_GPU_DEVICE (const ParticleType& p,
+                              amrex::Array4<int> const& partData,
+                              amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& plo_loc,
+                              amrex::GpuArray<amrex::Real,AMREX_SPACEDIM> const& dxi_loc)
+        {
+
+            auto iv = getParticleCell(p, plo_loc, dxi_loc, domain);
+            if(assign_buffer_grid(iv)==-1)
             {
-                virts.push_back(agg_particle.second);
+            //if this cell has no particle id info, do a straight copy
+            if(partData(iv,0)==0)
+            {
+          //                amrex::Gpu::Atomic::AddNoRet(&partData(iv,0), VirtualParticleID);
```

Comment:
Could you remove the commented-out line?

Reply:
Removed